File: tcg-java/test/de/kimminich/kata/tcg/GameBuilder.java
package de.kimminich.kata.tcg;

import static de.kimminich.kata.tcg.PlayerBuilder.anyPlayer;

public class GameBuilder {

    private Player activePlayer = anyPlayer();
    private Player opponentPlayer = anyPlayer();

    GameBuilder() {
    }

    public static GameBuilder aGame() {
        return new GameBuilder();
    }

    public static Game anyGame() {
        return aGame().build();
    }

    public Game build() {
        Game game = new Game(activePlayer, opponentPlayer);
        if (game.getActivePlayer() != activePlayer) {
            game.switchPlayer();
        }
        return game;
    }

    public GameBuilder withActivePlayer(Player activePlayer) {
        this.activePlayer = activePlayer;
        return this;
    }

    public GameBuilder withOpponentPlayer(Player opponentPlayer) {
        this.opponentPlayer = opponentPlayer;
        return this;
    }

}

File: tcg-kotlin/src/test/kotlin/de/kimminich/kata/tcg/GameBuilder.kt
package de.kimminich.kata.tcg

import org.mockito.kotlin.given
import org.mockito.kotlin.mock

class GameBuilder internal constructor() {
    private var activePlayer: Player? = null
    private var opponentPlayer: Player? = null
    private var startingPlayerChooser: Game.StartingPlayerChooser = mock()
    fun build(): Game {
        if(activePlayer == null) throw ExceptionInInitializerError("activePlayer must not be null")
        if(opponentPlayer == null) throw ExceptionInInitializerError("opponentPlayer must not be null")
        given(startingPlayerChooser.chooseBetween(activePlayer!!, opponentPlayer!!)).willReturn(activePlayer)
        return Game(activePlayer!!, opponentPlayer!!, startingPlayerChooser)
    }

    fun withActivePlayer(activePlayer: Player): GameBuilder {
        this.activePlayer = activePlayer
        return this
    }

    fun withOpponentPlayer(opponentPlayer: Player): GameBuilder {
        this.opponentPlayer = opponentPlayer
        return this
    }

    companion object {

        fun aGame(): GameBuilder {
            return GameBuilder()
        }


        fun anyGame(): Game {
            return  GameBuilder()
                .withActivePlayer(PlayerBuilder.anyPlayer())
                .withOpponentPlayer(PlayerBuilder.anyPlayer())
                .build()
        }
    }
}
